rfcs/0082-lib-experimental.md

```diff
@@ -0,0 +1,129 @@
+to tag libraries as explicitly being part of the *may not* category.
+
+Creating this experimental section to nixpkgs makes it clearer to maintainers
+which code they may use where, and which code they may not.
+If you are the maintainer of a package, and understand the risks related to using
+experimental code, you may use this code or accept patches using it.
+But you may also freely reject it, and no one should impose you that code without your consent.
+In particular you understand that the code and its API may change, and
+you may have to update your code accordingly, or accept patches that do.
+Thus you agree to collaborate with the author of the experimental code
+while the code remains in the experimental section.
+
+Code that should *not* go into `lib.experimental` but directly into `lib` include:
+* code that is good quality, well tested, known to work, with a stable API,
+  good documentation, and no controversy among maintainers of nixpkgs.
+
+Code that should *not* go into `lib.experimental` and not into nixpkgs at all include:
+* code that bad quality, untested, not working, etc.
+* code that is not yet used by any package in nixpkgs nor by any extension to nixpkgs.
+* code that while cool has no purpose in nixpkgs as such and can well live outside it.
+
+Making these social constraints explicit avoid accidents and conflict that would happen
+if they were left implicit.
+
+# Detailed design
+[design]: #detailed-design
+
+Experimental library code should be placed under the `lib/experimental` directory,
+where it will be made available in the `pkgs.lib.experimental` attrset,
+that will *not* re-export "direct" versions of the code directly in a parent attrset.
+
+A README.md will explain the purpose of this directory, by quoting this RFC and linking to it.
+
+The process for taking things out of experimental is when there is a consensus to do so:
+when enough maintainers use this code rather than alternatives or workarounds,
+when its API is stable and well documented, with enough years of good use,
+when multiple maintainers are ready to service the code and it's not controversial anymore,
+and/or when a RFC blesses the code to be made a direct part of lib.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+The [Gerbil](https://cons.io) packages in the
+http://github.com/muknio/nixpkgs/archive/devel.tar.gz fork of nixpkgs use a
+prototype object system called `POP` as a more versatile and principled alternative
+to the traditional extension mechanisms in `lib.fixedPoints` or `lib.customisation`.
+The libraries or use to good effects in particular by
+[Glow](http://gitlab.com/mukn/glow) for local overrides.
+
+Under this RFC, the code for POP would go in `lib/experimental/POP.nix`
+and be available as `pkgs.lib.POP`.
+Package maintainers may opt into using it like Gerbil does,
+but contributors may not use it without the consent
+of the maintainers of the code they contribute to.
```

Comment:
Something we should look into as well is how to enforce this. There's probably a nice way to do so using `scopedImport`.